lapack/gonum/dgghrd.go

```diff
@@ -10,57 +10,45 @@ import (
 )
 
-// Dgghrd reduces a pair of real matrices (A,B) to generalized upper
-// Hessenberg form using orthogonal transformations, where A is a
-// general matrix and B is upper triangular.  The form of the
-// generalized eigenvalue problem is
+// Dgghrd reduces a pair of real matrices (A,B) to generalized upper Hessenberg
+// form using orthogonal transformations, where A is a general matrix and B is
+// upper triangular.
 //
-//	A*x = lambda*B*x,
+// This subroutine simultaneously reduces A to a Hessenberg matrix H
 //
-// and B is typically made upper triangular by computing its QR
-// factorization and moving the orthogonal matrix Q to the left side
-// of the equation.
-// This subroutine simultaneously reduces A to a Hessenberg matrix H:
+//	Qᵀ*A*Z = H,
 //
-//	Qᵀ*A*Z = H
+// and transforms B to another upper triangular matrix T
 //
-// and transforms B to another upper triangular matrix T:
+//	Qᵀ*B*Z = T.
 //
-//	Qᵀ*B*Z = T
-//
-// in order to reduce the problem to its standard form
+// The orthogonal matrices Q and Z are determined as products of Givens
+// rotations. They may either be formed explicitly (lapack.OrthoExplicit), or
+// they may be postmultiplied into input matrices Q1 and Z1
+// (lapack.OrthoPostmul), so that
 //
-//	H*y = lambda*T*y
+//	Q1 * A * Z1ᵀ = (Q1*Q) * H * (Z1*Z)ᵀ,
+//	Q1 * B * Z1ᵀ = (Q1*Q) * T * (Z1*Z)ᵀ.
 //
-// where y = Zᵀ*x.
+// ilo and ihi determine the block of A that will be reduced. It must hold that
 //
-// The orthogonal matrices Q and Z are determined as products of Givens
-// rotations.  They may either be formed explicitly, or they may be
-// postmultiplied into input matrices Q1 and Z1, so that
+//   - 0 <= ilo <= ihi < n      if n > 0,
+//   - ilo == 0 and ihi == -1   if n == 0,
 //
-//	Q1 * A * Z1ᵀ = (Q1*Q) * H * (Z1*Z)ᵀ
-//	Q1 * B * Z1ᵀ = (Q1*Q) * T * (Z1*Z)ᵀ
-//
-// If Q1 is the orthogonal matrix from the QR factorization of B in the
-// original equation A*x = lambda*B*x, then Dgghrd reduces the original
-// problem to generalized Hessenberg form.
+// otherwise Dgghrd will panic.
 //
 // Dgghrd is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dgghrd(compq, compz lapack.OrthoComp, n, ilo, ihi int, a []float64, lda int, b []float64, ldb int, q []float64, ldq int, z []float64, ldz int) {
 	switch {
-	case compq != lapack.OrthoNone && compq != lapack.OrthoEntry && compq != lapack.OrthoUnit:
+	case compq != lapack.OrthoNone && compq != lapack.OrthoExplicit && compq != lapack.OrthoPostmul:
 		panic(badOrthoComp)
-	case compz != lapack.OrthoNone && compz != lapack.OrthoEntry && compz != lapack.OrthoUnit:
+	case compz != lapack.OrthoNone && compz != lapack.OrthoExplicit && compz != lapack.OrthoPostmul:
 		panic(badOrthoComp)
-	case len(a) < (n-1)*lda+n:
-		panic(shortA)
-	case len(b) < (n-1)*ldb+n:
-		panic(shortB)
 	case n < 0:
 		panic(nLT0)
-	case ilo < 0:
+	case ilo < 0 || max(0, n-1) < ilo:
 		panic(badIlo)
-	case ihi < ilo-1 || ihi >= n:
```

Comment:
Is `ihi >= n` not more readable since we are testing for ihi?

Reply:
The convention we use for ranges only uses `<`. We find that the expressions are clearer this way.